lib/Dialect/TTNN/IR/TTNNWorkaroundInterface.cpp

```diff
@@ -0,0 +1,136 @@
+
+      // Skip if the output result isn't a tensor
+      if (mlir::isa<::mlir::RankedTensorType>(operand.get().getType())) {
+        continue;
+      }
+
+      // Check if the output workaround is the same as the input DPS init
+      if (workarounds.getOutputOperandWorkarounds()[resultIndex] !=
+          workarounds.getInputOperandWorkarounds()[i]) {
+        return op->emitOpError() << "DPS output workaround does not match "
+                                    "the input DPS init operand workaround "
+                                 << i << " and " << resultIndex;
+      }
+
+      resultIndex++;
+    }
+  }
+
+  // All checks passed, return success
+  return mlir::success();
+}
+
+// Operand workarounds are defined for each operand and result of the operation.
+// If the operation is a DPS operation, same workarounds must be applied for the
+// DPS inits and DPS op outputs. All of this is verified in the interface
+// verifier. For example, if we have following ttnn operations:
+//
+// %0 = "ttnn.emptyOp"() : () -> tensor<1x1xf32>
+// %1 = "ttnn.abs"(%arg0, %0) : (tensor<1x1xf32>, tensor<1x1xf32>) ->
+// tensor<1x1xf32>
+//
+// In this example, we will have 2 input operand workarounds and 1 output
+// operand workaround, hence the output workaround must be the same as for the
+// second input operand.
+TTNNOperandsWorkarounds getDefaultOperandWorkarounds(Operation *op) {
+  // Special case empty op
+  // Empty op currently only supports creation in row major layout
+  if (mlir::isa<::mlir::tt::ttnn::EmptyOp>(op)) {
+    return WorkaroundFactory::createEmptyTTNNOperandsWorkarounds(0, 0)
+        .addOutputOperandWorkaround(
+            WorkaroundFactory::createEmptyTTNNOperandWorkarounds()
+                .setWorkaround(Layout::RowMajor));
+  }
+
+  if (mlir::dyn_cast<::mlir::tt::ttnn::AbsOp>(op)) {
+    return WorkaroundFactory::createEmptyTTNNOperandsWorkarounds(0, 0)
+        .addInputOperandWorkaround(
+            WorkaroundFactory::createEmptyTTNNOperandWorkarounds()
+                .setWorkaround(Layout::Tile))
+        .addInputOperandWorkaround(
+            WorkaroundFactory::createEmptyTTNNOperandWorkarounds()
+                .setWorkaround(Layout::Tile))
+        .addOutputOperandWorkaround(
+            WorkaroundFactory::createEmptyTTNNOperandWorkarounds()
+                .setWorkaround(Layout::Tile));
+  }
+
+  size_t tensorInputs = 0;
+  size_t tensorResults = 0;
+
+  // Count the number of tensor input operands including DPS inits
+  for (auto operand : op->getOperands()) {
+    if (mlir::isa<::mlir::RankedTensorType>(operand.getType())) {
+      tensorInputs++;
+    }
+  }
+
+  // Count the number of tensor results
+  for (auto result : op->getResults()) {
+    if (mlir::isa<::mlir::RankedTensorType>(result.getType())) {
+      tensorResults++;
+    }
+  }
+
+  return WorkaroundFactory::createEmptyTTNNOperandsWorkarounds(tensorInputs,
+                                                               tensorResults);
```

Comment:
Perhaps we should additionally have function:
```c++
Workarounds createEmptyTTNNOperandsWorkarounds(Operation* op);
```

This way you can put the loops above, counting tensor inputs/outputs, inside of the factory.

Reply:
Yeah, I agree, it is much more cleaner.